linalg/sparsemat.cpp

```diff
@@ -587,16 +616,66 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
    auto d_A = Read(A, nnz);
    auto d_x = x.Read();
    auto d_y = y.ReadWrite();
-   MFEM_FORALL(i, height,
+
+   //Skip if matrix has no non-zeros
+   if (nnz == 0) {return;}
+   if (Device::Allows(Backend::CUDA_MASK) && useCuSparse)
    {
-      double d = 0.0;
-      const int end = d_I[i+1];
-      for (int j = d_I[i]; j < end; j++)
+#ifdef MFEM_USE_CUDA
+      const double alpha = a;
+      const double beta  = 1.0;
+
+      //Initialize once
+      if (!initCuSparse)
       {
-         d += d_A[j] * d_x[d_J[j]];
+         /* create and setup matrix descriptor */
+         cusparseCreateCsr(&matA_descr,Height(), Width(), J.Capacity(),
+                           const_cast<int *>(d_I),
+                           const_cast<int *>(d_J), const_cast<double *>(d_A), CUSPARSE_INDEX_32I,
+                           CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, CUDA_R_64F);
+
+         /*Create handles for input/output vectors */
+         cusparseCreateDnVec(&vecX_descr, x.Size(), const_cast<double *>(d_x),
+                             CUDA_R_64F);
+         cusparseCreateDnVec(&vecY_descr, y.Size(), d_y, CUDA_R_64F);
+
+         /*Create allocate space for kernel */
+         cusparseSpMV_bufferSize(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, &alpha,
+                                 matA_descr,
+                                 vecX_descr, &beta, vecY_descr, CUDA_R_64F,
+                                 CUSPARSE_CSRMV_ALG1, &bufferSize);
+         if (bufferSize > 0)
+         {
+            cudaMalloc(&dBuffer, bufferSize);
+         }
+         initCuSparse = true;
```

Comment:
I think this code should be moved inside `initCuSparse()`; the design is a bit confusing currently: the method `initCuSparse()` does not do the initialization (and does not touch the `bool initCuSparse`) but instead the real initialization is done in `AddMult`, and when `if(!initCuSparse)` the code doesn't call `initCuSparse()` but does the "real" initialization.

Reply:
``InitCuSparse`` sets up the library for a given rank (only needs to be done once per MPI rank).  

 Indeed, the first call to ``AddMult``  allocates memory and wrappers for vectors and CSR matrices since a user could potentially setup a sparse matrix and **not** apply the action of it. In the case a user does  not do a mat-vec then I didn't see the need  to allocate buffers. Perhaps we can change the name of initCuSparse to initBuffers?

Reply:
I think that if we have only one buffer this logic would work.

Reply:
I'm not a big fan of the initialization being done in the `AddMult`, I see your point that sometime we want to create a `SparseMatrix` but we don't want to use it for matrix vector products, this seems a bit non standard to me, and if the buffer is shared anyway then it would not be as big of a problem.
I also prefer to have memory allocations (especially on GPU) not being done in the same function as `AddMult` because then when you profile it becomes harder to isolate these hidden costs. I think it's cleaner to have all the initialization costs to be in the constructor.